### ISSUE TITLE
Allow customization of enforcedNamespaceLabel behaviour

### ISSUE THREAD

**Issue by snowamk** [prometheus-operator/prometheus-operator#6775]
### Component(s)

Prometheus

### What is missing? Please describe.

Hello dear friends,

I would like to label all alerts automatically with the namespace that the PrometheusRule object was deployed in. This is to provide routing information to alerts that are coming from many different teams without requiring them to make any manual changes to their manifests. 

enforcedNamespaceLabel satisfies this requirement, except that it has extra functionality that breaks some other things, making it unfeasible to use. 

enforcedNamespaceLabel does four things currently:
```
When not empty, a label will be added to:  
1.  All metrics scraped from `ServiceMonitor`, `PodMonitor`, `Probe` and `ScrapeConfig` objects. 
2.  All metrics generated from recording rules defined in `PrometheusRule` objects. 
3.  All alerts generated from alerting rules defined in `PrometheusRule` objects. 
4.  All vector selectors of PromQL expressions defined in `PrometheusRule` objects. 
```
For my use case, only number 3 is required. 
Specifically, numbers 1 and 2 are unnecessary, and number 4 causes issues with rules that reference metrics deployed outside of the users namespace.

For example:
PrometheusRule in namespace `foo` has the following expression:
`absent(sum(kube_pod_status_ready{condition="true",pod=~".*redis-master.*"}))`
After setting `enforcedNamespaceLabel: workload_namespace`, this is automatically changed to
`absent(sum(kube_pod_status_ready{condition="true",pod=~".*redis-master.*",workload_namespace="foo"}))`

This then immediately fires, because kube-state-metrics and it's servicemonitor are deployed in `monitoring` namespace.
And in reality, `kube_pod_status_ready` has the label `workload_namespace="monitoring"`, not `"foo"`

Is the rule expression ideal? No. But the point is getting this behaviour without chasing up a bunch of different teams about their rules.

Ideally I would be able to say (the exact format doesn't matter)
```
enforcedNamespaceLabel:
  name: workload_namespace
  metrics: false 
  recordingRules: false
  alertingRules: true
  vectorSelectors: false
```
The granularity of `excludedFromEnforcement` does not allow customizing to this degree because disabling PrometheusRules also disables the behaviour I'm looking for. 

Thanks for reading.

### Describe alternatives you've considered.

1. Set the value to `namespace` instead of `workload_namespace` and worry about conflicts with the common namespace label. Especially behaviour number 1 overwriting kube-state-metrics' metrics with a hardcoded namespace label is a scary thought. The idea with using a label other than `namespace` was to avoid having to worry about conflicts with existing config. 
2. Modify downstream alert router to do the PrometheusRule lookup from the cluster manually and find out the namespace that way

## Environment
Kubernetes Version: 1.29
Prometheus-Operator Version: v0.74.0

**Comment by simonpasquier** [comment]
> And in reality, kube_pod_status_ready has the label workload_namespace="monitoring", not "foo"

Is there any label on kube_pod_status_ready that reports the "real" namespace where the pod is deployed?

**Comment by snowamk** [comment]
Yes, the label 'namespace'. See `Describe alternatives you've considered.` number 1

**Comment by alimehrabikoshki** [comment]
@simonpasquier did you get a chance to think about this one any further?

**Comment by alphabet5** [comment]
#4 is getting me currently. It'd be nice to not have metrics use the label so alert rules can use metrics from the ingress controller.

**Comment by hrobertson** [comment]
I would like the best of both worlds:

1. The convenience of automatically adding a namespace label to query expressions in rules (in case developers omit it and to facilitate deploying the same application (& PrometheusRule) in multiple namespaces without conflict).
2. The ability to not override a namespace label in a query expression if one has been explicitly set.

So rather than **enforcing** a namespace label on query expressions I would like to be able to default the namespace label if it's not set.

Anyone thinking about this issue should also keep in mind #6418 so that a minimally complex holistic solution can be reached.
